### PR TITLE
Added rename support for Qute templates

### LINKED ISSUE #492: Rename support
Today we have linked editing range support to rename an alias:
QuteLinkedEditingRangeDemo

but it should be nice to provide this same feature with `Rename (F2)`:

 * it will support another LSP client which doesn't support linked edting range.
 * it will support rename on referenced object which are declared outside the Qute template where rename is processed.

To provide this support, I suggest that we create a `QuteRename `class by copying/pasting https://github.com/redhat-developer/quarkus-ls/blob/master/qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteLinkedEditing.java and adapt it with rename.

### PR DESCRIPTION
**Pull request by AlexXuChen**
Added rename support for Qute templates
Peek 2022-06-07 18-48

Fixes #492 

Signed-off-by: Alexander Chen <alchen@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
You need to notify to the client that Qute LS can support rename. I suggest that you see LemMinx and you do the same thing for Qute LS, see 

 * https://github.com/eclipse/lemminx/blob/974099d749010f788f9ac7914894269e38f51a80/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/settings/capabilities/XMLCapabilityManager.java#L165
 * https://github.com/eclipse/lemminx/blob/974099d749010f788f9ac7914894269e38f51a80/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/settings/capabilities/ServerCapabilitiesInitializer.java#L56


After that you need to write tests. I suggest that you copy / paste assertRename in QuteAssert from LemMinx https://github.com/eclipse/lemminx/blob/974099d749010f788f9ac7914894269e38f51a80/org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/XMLAssert.java#L1357 and you adapt it for Qute.

I suggest that you create 3 tests class by pasting / copying tests from linked editing rangge https://github.com/redhat-developer/quarkus-ls/tree/master/qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/linkededitingrange

(ex : `com.redhat.qute.services.rename.QuteRenameInForSectionTest`) and your adapt it for rename

like we do for LemMinx https://github.com/eclipse/lemminx/blob/f9a795d895091fb2ac33b123980b61d86af3f94c/org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/XMLRenameTest.java

**Comment by AlexXuChen** [comment]
Still need to add tests

File: qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/QuteAssert.java [hunk 1]
```diff
@@ -787,14 +789,56 @@ public static void assertLocation(List<? extends Location> actual, Location... e
...
+			assertArrayEquals(expectedEdits.toArray(), Collections.emptyList().toArray());
+		}
+	}
+
+	public static List<TextEdit> edits(String newText, Range... ranges) {
+		List<TextEdit> re = new ArrayList<TextEdit>();
+		for (Range range : ranges) {
+			re.add(te(range.getStart().getLine(), range.getStart().getCharacter(), range.getEnd().getLine(),
+				range.getEnd().getCharacter(), newText));
+			// re.add(new TextEdit(range, newText))
```
Comment (AlexXuChen) [inline qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/QuteAssert.java:829]: Is the `te` helper preferred over creating a `TextEdit` instance, or vice versa?

Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/QuteAssert.java:829]: I suggest that you writ ethe method like this:

```
private static List<TextEdit> edits(String newText, Range... ranges) {
		return Stream.of(ranges).map(r -> new TextEdit(r, newText)).collect(Collectors.toList());
	}
```

File: qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/QuteAssert.java [hunk 2]
```diff
@@ -787,14 +789,56 @@ public static void assertLocation(List<? extends Location> actual, Location... e
...
+		QuteLanguageService languageService = new QuteLanguageService(new JavaDataModelCache(projectRegistry));
+
+		Position position = template.positionAt(offset);
+
+		WorkspaceEdit workspaceEdit = languageService.doRename(template, position, newText, () -> {
+		});
+		try {
+			List<TextEdit> actualEdits = workspaceEdit.getChanges().get(FILE_URI);
+			assertArrayEquals(expectedEdits.toArray(), actualEdits.toArray());
+		} catch (NullPointerException e) {
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/QuteAssert.java:819]: Wht you are catching NPE?

File: quarkus.jdt.ext/pom.xml [hunk 3]
```diff
@@ -38,7 +38,7 @@
     <repository>
       <id>jdt.ls.p2</id>
       <layout>p2</layout>
-      <url>https://download.eclipse.org/jdtls/snapshots/repository/latest/</url>
+      <url>https://download.eclipse.org/jdtls/milestones/1.12.0/repository/</url>
```
[the remark this answers is not in the document]
Comment (rgrunber) [reply quarkus.jdt.ext/pom.xml:41]: I can update this temporarily in lsp4mp + quarkus-ls to ensure we can build and test. However, we will need to do a proper update in both repositories to use Java 17 going forward.

[the remark this answers is not in the document]
Comment (angelozerr) [reply quarkus.jdt.ext/pom.xml:41]: Ok but perhaps we could do that in an another PR?

File: quarkus.jdt.ext/pom.xml [hunk 4]
```diff
@@ -263,5 +263,5 @@
     <module>com.redhat.microprofile.jdt.quarkus.test</module>
     <module>com.redhat.microprofile.jdt.quarkus.site</module>
   </modules>
-  
+
```
Comment (angelozerr) [inline quarkus.jdt.ext/pom.xml:266]: Please don't update this file.

File: qute.jdt/pom.xml [hunk 5]
```diff
@@ -101,7 +101,7 @@
 					<configuration>
 						<useUIHarness>false</useUIHarness>
 						<argLine>${tycho.test.jvmArgs}</argLine>
-						<!-- kill test JVM if tests take more than 1 minute (60 seconds) to 
+						<!-- kill test JVM if tests take more than 1 minute (60 seconds) to
```
Comment (angelozerr) [inline qute.jdt/pom.xml:104]: Please don't update this file.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/template/TemplateFileTextDocumentService.java [hunk 6]
```diff
@@ -84,14 +86,14 @@ public TemplateFileTextDocumentService(QuteLanguageServer quteLanguageServer, Sh
 	public void didOpen(DidOpenTextDocumentParams params) {
 		QuteTextDocument document = (QuteTextDocument) documents.onDidOpenTextDocument(params);
 		document.getProjectInfoFuture() //
-				.thenAccept(projectInfo -> {
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/template/TemplateFileTextDocumentService.java:87]: Could you please don't update the format of the code in this file. It is hard to review it. The main problem is that your formatter (from vscode-java) seems not have the same behavior than Eclipse IDE formatter.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/template/TemplateFileTextDocumentService.java [hunk 7]
```diff
@@ -204,32 +205,40 @@ public CompletableFuture<List<DocumentLink>> documentLink(DocumentLinkParams par
...
 	public CompletableFuture<List<? extends Location>> references(ReferenceParams params) {
 		return getTemplate(params.getTextDocument(), (cancelChecker, template) -> {
 			return getQuteLanguageService().findReferences(template, params.getPosition(), params.getContext(),
-					cancelChecker);
+				cancelChecker);
+		});
+	}
+
+	@Override
+	public CompletableFuture<WorkspaceEdit> rename(RenameParams params) {
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/template/TemplateFileTextDocumentService.java:238]: In other words, you should just add this piece of code (other code should not be updated just for formatting).

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteCodeActions.java [hunk 8]
```diff
@@ -81,12 +81,12 @@ class QuteCodeActions {
 	private static final String UNDEFINED_NAMESPACE_SEVERITY_SETTING = "qute.validation.undefinedNamespace.severity";
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteCodeActions.java:81]: Please don't update this file. Your PR is about rename and not for codeAction.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteRename.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
+/*******************************************************************************
+* Copyright (c) 2021 Red Hat Inc. and others.
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteRename.java:2]: 2022

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteRename.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+				textEdits.add(new TextEdit(r, newText));
+			}
+			return createWorkspaceEdit(template.getUri(), textEdits);
+		} catch (BadLocationException e) {
+			LOGGER.log(Level.SEVERE, "In QuteRename, the client provided Position is at a BadLocation", e);
+			return null;
+		}
+	}
+
+	private WorkspaceEdit createWorkspaceEdit(String documentURI, List<TextEdit> textEdits) {
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteRename.java:67]: private static...

File: qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/QuteAssert.java [hunk 11]
```diff
@@ -113,47 +114,48 @@ public static void testCompletionFor(String value, CompletionItem... expectedIte
 	}
 
 	public static void testCompletionFor(String value, String fileUri, CompletionItem... expectedItems)
-			throws Exception {
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/QuteAssert.java:116]: Please don't format the code of QuteAssert (only for your new method). It is hard to review it.

File: qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/QuteAssert.java [hunk 12]
```diff
@@ -787,14 +789,52 @@ public static void assertLocation(List<? extends Location> actual, Location... e
 		assertArrayEquals(expected, actual.toArray());
 	}
 
+	// ------------------- Rename assert
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/QuteAssert.java:792]: In other words, you should just adding thi spiece of code.

**Comment by angelozerr** [comment]
Great job @AlexXuChen !

**Pull request merged**